server/schedulers/evict_slow_store.go

```diff
@@ -0,0 +1,221 @@
+	return EvictSlowStoreName
+}
+
+type evictSlowStoreScheduler struct {
+	*BaseScheduler
+	conf                 *evictSlowStoreSchedulerConfig
+	evictLeaderScheduler *evictLeaderScheduler
+}
+
+func (s *evictSlowStoreScheduler) GetName() string {
+	return EvictSlowStoreName
+}
+
+func (s *evictSlowStoreScheduler) GetType() string {
+	return EvictSlowStoreType
+}
+
+func (s *evictSlowStoreScheduler) EncodeConfig() ([]byte, error) {
+	return schedule.EncodeConfig(s.conf)
+}
+
+func (s *evictSlowStoreScheduler) Prepare(cluster opt.Cluster) error {
+	if s.evictLeaderScheduler != nil {
+		return s.prepareEvictLeader(cluster)
+	}
+	return nil
+}
+
+func (s *evictSlowStoreScheduler) prepareEvictLeader(cluster opt.Cluster) error {
+	var res error
+	for id := range s.evictLeaderScheduler.conf.StoreIDWithRanges {
+		if err := cluster.SlowStoreEvicted(id); err != nil {
+			res = err
+		}
+	}
+	return res
+}
+
+func (s *evictSlowStoreScheduler) Cleanup(cluster opt.Cluster) {
+	if s.evictLeaderScheduler != nil {
+		s.cleanupEvictLeader(cluster)
+	}
+}
+
+func (s *evictSlowStoreScheduler) cleanupEvictLeader(cluster opt.Cluster) {
+	for id := range s.evictLeaderScheduler.conf.StoreIDWithRanges {
+		cluster.SlowStoreRecovered(id)
+	}
+}
+
+func (s *evictSlowStoreScheduler) IsScheduleAllowed(cluster opt.Cluster) bool {
+	if s.evictLeaderScheduler != nil {
+		return s.evictLeaderScheduler.IsScheduleAllowed(cluster)
+	}
+	return true
+}
+
+func (s *evictSlowStoreScheduler) Schedule(cluster opt.Cluster) []*operator.Operator {
+	schedulerCounter.WithLabelValues(s.GetName(), "schedule").Inc()
+	var ops []*operator.Operator
+
+	evictedStores := s.conf.EvictedStores
+	if len(evictedStores) != 0 {
+		store := cluster.GetStore(evictedStores[0])
+		if store == nil || store.IsTombstone() {
+			// Previous slow store had been removed, remove the sheduler and check
+			// slow node next time.
+			log.Info("slow store has been removed",
+				zap.Stringer("store", store.GetMeta()))
+		} else if store.GetSlowScore() <= slowStoreRecoverThreshold {
+			log.Info("slow store has been recovered",
+				zap.Stringer("store", store.GetMeta()))
+		} else {
+			return s.evictLeaderScheduler.Schedule(cluster)
+		}
+		// Stop to evcit leaders
+		s.conf.EvictedStores = []uint64{0}
+		s.conf.Persist()
+		s.cleanupEvictLeader(cluster)
+		s.evictLeaderScheduler = nil
+	} else {
+		slowStores := make([]*core.StoreInfo, 0)
+		for _, store := range cluster.GetStores() {
+			if store.IsTombstone() {
+				continue
+			}
+
+			if store.IsUp() && store.IsSlow() {
+				slowStores = append(slowStores, store)
+			}
+		}
+
+		// If there is only one slow store, evict leaders from that store.
+		if len(slowStores) == 1 && slowStores[0].GetSlowScore() >= slowStoreEvictThreshold {
+			store := slowStores[0]
+			log.Info("detected slow store, start to evict leaders",
+				zap.Stringer("store", store.GetMeta()))
+			s.conf.EvictedStores = []uint64{store.GetID()}
```

Comment:
As we can evict leader from only 1 slow store, why use a vector to store the ID>

Reply:
I thought there would be a chance to introduce a more complex strategy for this scheduler. So use a vector here for keeping flexibility.